File: Difficulty/Medium/[166] Fraction to Recurring Decimal/solution.py
class Solution:
    def fractionToDecimal(self, numerator: int, denominator: int) -> str:
        if numerator == 0:
            return "0"
        res = ''
        if (numerator < 0) ^ (denominator < 0):
            res += '-'
        n, d = abs(numerator), abs(denominator)
        res += str(n // d)
        if n % d == 0:
            return res
        res += '.'
        m = dict()
        r = n % d
        while r:
            if m.get(r, None):
                res = res[:m[r]] + '(' + res[m[r]:] + ')'
                break
            m[r] = len(res)
            r *= 10
            res += str(r // d)
            r %= d
        return res

File: Difficulty/Medium/[166] Fraction to Recurring Decimal/solution.cpp
/*
 * @lc app=leetcode id=166 lang=cpp
 *
 * [166] Fraction to Recurring Decimal
 */

// @lc code=start
class Solution {
public:
    string fractionToDecimal(int numerator, int denominator) {
        if (numerator == 0) return "0";
        string res;
        if (numerator < 0 ^ denominator < 0) res += '-';
        int64_t n = abs(int64_t(numerator));
        int64_t d = abs(int64_t(denominator));
        res += to_string(n / d);
        if (n % d == 0) return res;
        res += '.';
        unordered_map<int, int> map;
        for (int64_t r = n % d; r; r %= d) {
            if (map.count(r) > 0) {
                res.insert(map[r], 1, '(');
                res += ')';
                break;
            }
            map[r] = res.size();
            r *= 10;
            res += to_string(r / d);
        }
        return res;
    }
};
// @lc code=end

